python/tvm/contrib/hexagon/build.py

```diff
@@ -386,21 +384,46 @@ def _run_server_script(self):
             stderr=subprocess.PIPE,
         )
 
+    def _cleanup_port_forwarding(self):
+        # Removed pre-defined forward/reverse rules
+        rpc_tracker_port = self._rpc_info["rpc_tracker_port"]
+        subprocess.check_call(
+            self._adb_device_sub_cmd + ["reverse", "--remove", f"tcp:{rpc_tracker_port}"]
+        )
+        for port in self.forwarded_ports_:
+            subprocess.check_call(self._adb_device_sub_cmd + ["forward", "--remove", f"tcp:{port}"])
+
+    def _terminate_remote(self):
+        # Send interupt to main and child processes
+        SIGINT = 2
+        subprocess.Popen(
+            self._adb_device_sub_cmd
+            + ["shell", f"pkill -l{SIGINT} -P `cat {self._workspace}/rpc_pid.txt`"]
+        )
+        subprocess.Popen(
+            self._adb_device_sub_cmd
+            + ["shell", f"kill -s sigint `cat {self._workspace}/rpc_pid.txt`"]
+        )
+        subprocess.Popen(
+            self._adb_device_sub_cmd
+            + ["shell", f"kill -s sigint `cat {self._workspace}/rpc_pid.txt`"]
+        )
+        # Wait for processes to destruct cleanly after receiving the intrupt
+        subprocess.Popen(self._adb_device_sub_cmd + ["shell", "sleep", "0.1s"])
+        # Kill process children
+        self._adb_device_sub_cmd + ["shell", f"pkill -P `cat {self._workspace}/rpc_pid.txt`"]
```

Comment:
I wonder if we could set the process group for the subprocesses launched by `PopenWorker`.  Those are the ones that tend to be left-over.  The `pkill -P` command would send signals to the direct children of the main process, but not to any grandchildren (e.g. main process starts a worker subprocess, which starts a compilation subprocess), or to any processes started after the child process is killed but before the parent process is killed (e.g. main process interprets the closing child as a failed test case, starts a new test case).  It [looks like](https://pymotw.com/2/subprocess/#process-groups-sessions) it would be relatively straightfoward to have `Popen` calls started as the same process group, so that they would all receive the signal simultaneously.